istioctl/cmd/istiod-list.go

```diff
@@ -0,0 +1,312 @@
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"istio.io/istio/operator/cmd/mesh"
+	operator_istio "istio.io/istio/operator/pkg/apis/istio"
+	iopv1alpha1 "istio.io/istio/operator/pkg/apis/istio/v1alpha1"
+	"istio.io/istio/operator/pkg/manifest"
+	"istio.io/istio/operator/pkg/util"
+	"istio.io/istio/operator/pkg/util/clog"
+	"istio.io/istio/pkg/config"
+)
+
+type istiodListArgs struct {
+	// manifestsPath is a path to a charts and profiles directory in the local filesystem, or URL with a release tgz.
+	manifestsPath string
+}
+
+var (
+	istioOperatorGVR = apimachinery_schema.GroupVersionResource{
+		Group:    iopv1alpha1.SchemeGroupVersion.Group,
+		Version:  iopv1alpha1.SchemeGroupVersion.Version,
+		Resource: "istiooperators",
+	}
+)
+
+func istiodListCmd() *cobra.Command {
+	kubeConfigFlags := &genericclioptions.ConfigFlags{
+		Context:    strPtr(""),
+		Namespace:  strPtr(""),
+		KubeConfig: strPtr(""),
+	}
+	listArgs := &istiodListArgs{}
+
+	listCmd := &cobra.Command{
+		Use:   "list",
+		Short: "Lists Istio control planes",
+		Long:  "The list subcommand displays installed Istio control planes to the console",
+		// nolint: lll
+		Example: `  # List Istio installations
+  istioctl experimental istiod list
+
+`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			l := clog.NewConsoleLogger(cmd.OutOrStdout(), cmd.ErrOrStderr(), scope)
+			return istiodList(cmd.OutOrStdout(), listArgs, kubeConfigFlags, l)
+		}}
+
+	listCmd.PersistentFlags().StringVarP(&listArgs.manifestsPath, "manifests", "d", "", mesh.ManifestsFlagHelpStr)
+	return listCmd
+}
+
+func istiodList(writer io.Writer, listArgs *istiodListArgs, restClientGetter genericclioptions.RESTClientGetter, l clog.Logger) error {
+	restConfig, err := restClientGetter.ToRESTConfig()
+	if err != nil {
+		return err
+	}
+	iops, err := getIOPs(restConfig)
+	if err != nil {
+		return err
+	}
+	return printIOPs(writer, iops, listArgs.manifestsPath, restConfig, l)
+}
+
+func printIOPs(writer io.Writer, iops []*iopv1alpha1.IstioOperator, manifestsPath string, restConfig *rest.Config, l clog.Logger) error {
+	if len(iops) == 0 {
+		_, err := fmt.Fprintf(writer, "No IstioOperators present.\n")
+		return err
+	}
+
+	sort.Slice(iops, func(i, j int) bool {
+		return iops[i].Spec.Revision < iops[j].Spec.Revision
+	})
+
+	w := new(tabwriter.Writer).Init(writer, 0, 8, 1, ' ', 0)
+	fmt.Fprintln(w, "VERSION\tREVISION\tPROFILE\tCUSTOMIZATIONS\tPODS\tAGE")
+	for _, iop := range iops {
+		podCount, deploymentAge, err := getControlPlaneDeployment(iop, manifestsPath, restConfig)
+		if err != nil {
+			podCount = "<error>"
+			deploymentAge = "<error>"
+		}
+
+		diffs, err := getDiffs(iop, manifestsPath, effectiveProfile(iop.Spec.Profile), l)
+		if err != nil {
+			return err
+		}
+		for i, diff := range diffs {
+			if i == 0 {
+				fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\t%s\n",
+					iop.Spec.Tag,
+					renderWithDefault(iop.Spec.Revision, "master"),
```

Comment:
This should also be `default` instead of `master`?

Reply:
I think it would be better to use an invalid name like `<default>` when no revision is present. This is because the user could potentially use `default` or `master` and this could be confusing.